pkg/controller/registry/resolver/step_resolver.go

```diff
@@ -85,13 +113,40 @@ func NewOperatorStepResolver(lister operatorlister.OperatorLister, client versio
 	return stepResolver
 }
 
+func (r *OperatorStepResolver) WithGlobalNamespaceToggle(globalNamespaceToggle GlobalNamespaceToggle) *OperatorStepResolver {
+	r.globalCatalogSourceToggle = globalNamespaceToggle
+	return r
+}
+
 func (r *OperatorStepResolver) ResolveSteps(namespace string) ([]*v1alpha1.Step, []v1alpha1.BundleLookup, []*v1alpha1.Subscription, error) {
 	subs, err := r.listSubscriptions(namespace)
 	if err != nil {
 		return nil, nil, nil, err
 	}
 
-	namespaces := []string{namespace, r.globalCatalogNamespace}
+	includeGlobalCatalogs, err := func() (bool, error) {
+		if r.globalCatalogSourceToggle != nil {
+			toggle, err := r.globalCatalogSourceToggle.ShouldEnableGlobalNamespaceResolution(context.Background(), namespace)
+			if err != nil {
+				return false, err
+			}
+			return toggle, nil
+		}
+		return true, nil
+	}()
+
+	fmt.Printf("Disabling Global Catalog Sources: %t\n", includeGlobalCatalogs)
+
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	namespaces := []string{namespace}
+
+	if includeGlobalCatalogs {
+		namespaces = append(namespaces, r.globalCatalogNamespace)
+	}
+
```

Comment:
This kind of behavior is exactly what cache.SourceProvider is meant to implement. It is effectively `f(namespaces) -> sources`, so you should be able to write an implementation that conditionally removes the global namespace and then delegates to another SourceProvider.